server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileUtils.java

```diff
@@ -0,0 +1,195 @@
+     */
+    private static final int ZOOM_SHIFT = MAX_ZOOM * 2;
+
+    /**
+     * Bit mask to extract just the lowest 29 bits of a long
+     */
+    private static final long X_Y_VALUE_MASK = (1L << MAX_ZOOM) - 1;
+
+    /**
+     * Parse an integer precision (zoom level). The {@link ValueType#INT} allows it to be a number or a string.
+     *
+     * The precision is expressed as a zoom level between 0 and {@link #MAX_ZOOM} (inclusive).
+     *
+     * @param parser {@link XContentParser} to parse the value from
+     * @return int representing precision
+     */
+    static int parsePrecision(XContentParser parser) throws IOException, ElasticsearchParseException {
+        final Object node = parser.currentToken().equals(XContentParser.Token.VALUE_NUMBER)
+            ? Integer.valueOf(parser.intValue())
+            : parser.text();
+        return XContentMapValues.nodeIntegerValue(node);
+    }
+
+    /**
+     * Assert the precision value is within the allowed range, and return it if ok, or throw.
+     */
+    static int checkPrecisionRange(int precision) {
+        if (precision < 0 || precision > MAX_ZOOM) {
+            throw new IllegalArgumentException("Invalid geotile_grid precision of " +
+                precision + ". Must be between 0 and " + MAX_ZOOM + ".");
+        }
+        return precision;
+    }
+
+    /**
+     * Encode lon/lat to the geotile based long format.
+     * The resulting hash contains interleaved tile X and Y coordinates.
+     * The precision itself is also encoded as a few high bits.
+     */
+    static long longEncode(double longitude, double latitude, int precision) {
+        // Mathematics for this code was adapted from https://wiki.openstreetmap.org/wiki/Slippy_map_tilenames#Java
+
+        // Number of tiles for the current zoom level along the X and Y axis
+        final int tiles = 1 << checkPrecisionRange(precision);
+
+        int xTile = (int) Math.floor((normalizeLon(longitude) + 180) / 360 * tiles);
+
+        double latSin = Math.sin(Math.toRadians(normalizeLat(latitude)));
+        int yTile = (int) Math.floor((0.5 - (Math.log((1 + latSin) / (1 - latSin)) / (4 * Math.PI))) * tiles);
+
+        // Edge values may generate invalid values, and need to be clipped.
+        // For example, polar regions (above/below lat 85.05112878) get normalized.
+        if (xTile < 0) {
+            xTile = 0;
+        }
+        if (xTile >= tiles) {
+            xTile = tiles - 1;
+        }
+        if (yTile < 0) {
+            yTile = 0;
+        }
+        if (yTile >= tiles) {
+            yTile = tiles - 1;
+        }
+
+        // Zoom value is placed in front of all the bits used for the geotile
+        // e.g. when max zoom is 29, the largest index would use 58 bits (57th..0th),
+        // leaving 5 bits unused for zoom. See MAX_ZOOM comment above.
+        return ((long) precision << ZOOM_SHIFT) | ((long) xTile << MAX_ZOOM) | ((long) yTile);
```

Comment:
Should we just make all the relevant variables/static defaults `long` instead of `int`, to save on all the casting?  Not because it matters for performance, but just for readability.

I don't think we're relying on int overflows anywhere, so it should be safe to just use longs?

I don't care strongly, so feel free to ignore this (or if I missed somewhere else that needs them as ints).  Was just something I noticed in passing :)